### PR TITLE
feat(python): Add pyarrow write_to_dataset to write_parquet function

### PR DESCRIPTION
**Pull request by potzenhotz**
Implements [#9812](https://github.com/pola-rs/polars/issues/9812)

Currently Polars uses _pyarrow.write_table()_ to write parquets with pyarrow. The _pyarrow.write_to_dataset()_ funtion is not used. 
_pyarrow.write_to_dataset()_ does support more features than _write_table()_. Important features like writing partitioned parquet files.
Therefore, I added _pyarrow.write_to_dataset()_ using a new parameter in write_parquet(), namely "use_pyarrow_write_to_dataset".

I added two tests and an example in the docstring.
Currently some parameters like statistics or compressions are not used when calling _write_to_dataset()_ as these are not supported in _write_to_dataset()_.

Links:

- https://arrow.apache.org/docs/python/generated/pyarrow.parquet.write_to_dataset.html

### CODE REVIEW

**Comment by alexander-beedie** [comment]
Hmm… If this function represents a superset of `write_table` then we should consider replacing the existing call instead of supporting both and adding an extra parameter.

I’m on a train at the moment, but reading through the linked docstring it seems like the `**kwargs` allow pass-through options to the underlying parquet writes that would include compression/etc?

**Comment by potzenhotz** [comment]
Hey @alexander-beedie,

so this was my first idea using use_legacy_dataset=True as default so that write_table is used and setting use_legacy_dataset=False when pyarrow.write_to_dataset should be used.
But, we get a dprecation warning from pyarrow as it is not supported in the future (https://github.com/apache/arrow/issues/32067) 
`FutureWarning: Passing 'use_legacy_dataset=True' to get the legacy behaviour is deprecated as of pyarrow 11.0.0, and the legacy implementation will be removed in a future version.`

Therefore, I left write_table() unchanged.

I added the compression, statistic etc. fields to the to pyarrow_options and edited one of the test cases.

Do we need further tests or code changes? 
Thanks for the fast response already.

**Comment by erinov1** [comment]
Even in the case of no partitioning columns, `write_to_dataset` creates a data directory with a parquet file named `<uuid>-0.parquet`. In other words, `df.write_parquet("test.parquet", use_pyarrow_write_to_dataset=True)` would write a directory while `df.write_parquet("test.parquet", use_pyarrow_write_to_dataset=False)` would write a single file. If this proposal is adopted then the docs should reflect the difference.

**Comment by potzenhotz** [comment]
Hey, thats true but the same behaviour as Spark which is a almost standard in Cloud ETL.
I added two more lines in the description.

**Comment by alexander-beedie** [comment]
Ok; let's see how this looks instead - rather than have another (long!) parameter called `use_pyarrow_write_to_dataset`, it could be a lot cleaner to automatically use `write_to_dataset` _if_ the user supplies "partition_cols" to the pyarrow options dict (obviously keeping/tweaking the docs to make it clear that this is what happens). It keeps the simplicity of a single param and it feels fairly intuitive - the caller doesn't need to switch between different functions themselves as their given options should imply which one should be used 🤔

**Comment by potzenhotz** [comment]
Fair argument. I changed the code accordingly.

**Comment by potzenhotz** [comment]
Hey @alexander-beedie is there anything futher to change? Are there any plans to integrate te change?

Thanks.

**Comment by alexander-beedie** [comment]
> Hey @alexander-beedie is there anything futher to change? Are there any plans to integrate te change?

Ahh, great - I didn't spot that you had made the update - will take a look :)

**Review by alexander-beedie — CHANGES_REQUESTED** [review]
If you can update `test_write_parquet_using_pyarrow_write_to_dataset_with_partitioning` as requested, should be good to go 👍

**Comment by uditrana** [comment]
Have been looking for this feature! Hope to see it added soon.

File: py-polars/tests/unit/io/test_parquet.py [hunk 1]
```diff
@@ -46,6 +46,22 @@ def test_write_parquet_using_pyarrow_9753(tmpdir: Path) -> None:
     )
 
 
+@pytest.mark.parametrize("compression", COMPRESSIONS)
+def test_write_parquet_using_pyarrow_write_to_dataset_with_partitioning(
```
[the remark this answers is not in the document]
Comment (potzenhotz) [reply py-polars/tests/unit/io/test_parquet.py:50]: Enhanced the test case by reading the written parquets and asserted them.

**Review by alexander-beedie — APPROVED** [review]

**Comment by alexander-beedie** [comment]
Thanks!

**Comment by potzenhotz** [comment]
Thanks for the great feedback and communication!

**Pull request merged**